### ISSUE TITLE
Migration to a new hosting

### ISSUE THREAD

**Issue by neuoy** [neuoy/KSPTrajectories#150]
I'm considering to move Trajectories away from Github. I was already reluctant to host it there in the first place, since it's basically out of anyone's control, could disappear anytime, lock data away, require payment, etc. This gets confirmed by the fact it's going to be bought by Microsoft (Did anyone ask me if I was fine with this idea? Well, no, why would users have any voice since they were never given any control in the first place?)

So... Not a big deal, but I'd still like to move.

I'm currently evaluating self-hosting (gitlab and gitea), which are promising solutions. I'd like some form or replication of the whole project (code and metadata, including issues) so that it would not go down if something bad happens (to myself, my will to host, or anything). A simple solution would be to publish backups somewhere "safe", so that someone can come up with a new instance if needed.

I'm open to other propositions though. I'm also open to let the project on github if you want to stay there. After all I'm not active on this project for quite some time now.

In case the project moves, the master branch would be deleted on github, and the main branch would be set to a new one, containing only a readme.md indicating the new location of the project. Of course links would be updated on the KSP forum.

What do you think?

**Comment by Gordon-Dry** [comment]
Well, afaik did MS use a lot of Github over the last years, this was just a logical step.

But who can decide if it's better to be careful (aka. paranoid) or if it's too much of a hassle, just being MS hater.

The one extreme or the other, or something inbetween.

**Comment by fat-lobyte** [comment]
> since it's basically out of anyone's control, could disappear anytime, lock data away, require payment, 

I'm not really worried about these things. The risk that this happens is fundamental to hosting it literally anywhere. Our big advantage, however, is that we are free software that uses git. If it disappears, we move somewhere else. If they lock data away, we move somewhere else. If they require payment, ... you get the idea. In case they tamper with our history, we see it immediately in the next pull request.

The code is safe, because dozens of people have it lying around on their machines. The issues and accompanying discussions are a bit of a different story, but maybe we can extract those on a regular basis.

> This gets confirmed by the fact it's going to be bought by Microsoft

Again, I'm not particularly worried. In addition to the stuff mentioned above, I don't see much cause for action because:

* They announced no significant changes to what they plan to do
* They claim that they want to remain open and free for everyone
* MS's business strategy in the last few years seem to align with that

So what this boils down to is trust. Do you trust Microsoft? Most people seem not to trust them, and that's understandable, they've done some shady shit in the past.

Personally, I will give them the benefit of the doubt (until the first shady thing happens), because I think they have started to realize that you can make a lot of money with open source, even if you don't fuck people over. So IMO, keeping GitHub free and open is in their business interest.

But that's just my opinion and I absolutely understand if peoples opinions differ.

> I'm currently evaluating self-hosting (gitlab and gitea), which are promising solutions.

I would very, very much prefer to not do self-hosting.

First off, it's extra work for somebody who has to make sure and do maintenance tasks on a regular basis. That's a big problem for volunteers, since volunteers tend to disappear (I'm looking at myself, I am guilty of this) and things don't get maintained for months. Self-hosting would need regular updates (at least for security).

Second, hosting costs money, and that does have to come from somewhere. I'm not particularly inclined to solicit donations for hosting, and I'm also not particularly inclined to just let somebody (even if it's one of us maintainers) host it on a server in someones home.

Third, I believe that it's important for an open source project to lower the barrier to contributions as much as possible. Creating accounts on strange servers just to contribute code is not a thing I do with pleasure, and it doesn't exactly increase visibility. There are millions of people on GitHub and a lot (but fewer) on GitLab, who already have an account there. Tracking forks is much easier when most people will have an account there anyway, and we are searchable so that people can find us.

Fourth, a self-hosted copy prohibits cross-linking and public communication with other mod projects, for example with kOS.

Therefor, my suggestion is to either stay on GitHub or move to GitLab - the original hosted version, and not a self-hosted copy. My vote is 50:50 between those two options.

> I'd like some form or replication of the whole project (code and metadata, including issues) so that it would not go down if something bad happens (to myself, my will to host, or anything).

We could do that regardless of where and whether we move.

> In case the project moves, the master branch would be deleted on github, and the main branch would be set to a new one, containing only a readme.md indicating the new location of the project.

I don't think there is any reason to do that other than to spite Microsoft. Since we are an open source program, there is absolutely no harm in just leaving the code here, kind of as a backup. We can disable issues and pull requests, and set the default branch to a branch that you mentioned, but not necessarily delete the master and other branches.

**tl; dr, I think we should either stay or move to the (main) GitLab, but not self-host.**

**Comment by neuoy** [comment]
These are sensible arguments, which make sense, thanks for sharing them.

I'll try to clarify my thoughts (which are still a bit fuzzy) :

- I don't have much against Microsoft specifically (most I have comes from [here](http://sebsauvage.net/wiki/doku.php?id=microsoft) and my personal experience with Windows 10, Visual Studio, and Skype, but this is mostly out of the subject at hand)
- I do have much against big companies in general (too much power in too few hands, and I'm also against capitalism in general, but that's again out of the subject)
- I don't like the idea of concentrating more knowledge (about softwares, their developers, and their users) in Microsoft hands, which are already collecting lot of data with Windows 10, Azure, and other products (that would be the same issue with Google, or Amazon, etc.)
- Microsoft obviously (in my opinion) is bullshiting people about their motivations ; they want to lock people in their services (azure, github, visual studio), make them dependent, for profit, not to empower them. At best, these motivations (profit and empowerment) are compatible, but still, they are lying, in my opinion (some would say "making marketing communication" but I think this is pretty much the same thing)
- lastly, Microsoft has some history of dumping source control platforms (codeplex) ; by the way they encouraged users to migrate on github, which is funny (at best) ; but there is no reason for that to happen again in the short term.

When I said "This gets confirmed by the fact it's going to be bought by Microsoft" I was thinking about the lack of control of users on github. For the reasons stated above, I don't like the idea of Microsoft taking control of github, which highlights the fact that I had no voice in the matter. This problem is a github problem, not related to Microsoft. It's similar for the hosted version of gitlab.

I like the idea of migrating to gitlab as well ; as I said I'm open to a lot of options. I'm fine to let the project on github as well. I'll migrate my other projects, but Trajectories is clearly not *my* project anymore, so if I'm the only one to want to move it, there's no reason to do it (and I would honestly be fine with that, no pressure).

You are right about keeping the master branch as a backup, there is even an option in github to archive a project (make it read-only), if it comes to that.

**Comment by neuoy** [comment]
Also, I think Gitlab community was considering the idea of making a federation system (similar to Mastodon), which would solve some of the issues (people could comment, fork and send pull requests on any instance, without creating a new account). But it might never be a thing.

**Comment by PiezPiedPy** [comment]
After using GitLab for a few months I feel that GitHub is a much better platform, I did not like using GitLab, also end users on the KSP forums are more used to seeing mods on GitHub.

MS did state that they would be keeping GitHub open source and would not be making any fundamental changes, time will tell.

I myself do not like the idea of MS taking over GitHub, although I feel it would be best to stay with GitHub until MS decide to make a mess of things, i.e. they assume we all need a MS account to log into GitHub or something just as stupid.

**Comment by neuoy** [comment]
OK, that settles it, Trajectories stays on github for now :-) Thanks for your feedback.
